lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,205 @@
+      assign(ctx,
+        fields: fields,
+        options: %{},
+        vl_alias: nil,
+        missing_dep: missing_dep(),
+        fresh: attrs == %{}
+      )
+
+    {:ok, ctx, reevaluate_on_change: true}
+  end
+
+  @impl true
+  def scan_binding(pid, binding, env) do
+    data_options =
+      for {key, val} <- binding, is_atom(key), is_map(val), into: %{}, do: {key, Map.keys(val)}
+
+    vl_alias = vl_alias(env)
+    send(pid, {:scan_binding_result, data_options, vl_alias})
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      fields: ctx.assigns.fields,
+      missing_dep: ctx.assigns.missing_dep,
+      options: ctx.assigns.options,
+      fresh: ctx.assigns.fresh
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, data_options, vl_alias}, ctx) do
+    ctx = assign(ctx, options: data_options, vl_alias: vl_alias)
+    broadcast_event(ctx, "set_available_data", %{"options" => data_options})
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => field, "value" => value}, ctx) do
+    current_data = ctx.assigns.fields["data_variable"]
+    current_field = ctx.assigns.fields[field]
+
+    updated_fields = to_update(field, value)
+    ctx = update(ctx, :fields, &Map.merge(&1, updated_fields))
+
+    if field == "data_variable" && value != current_data, do: update_options(ctx, value)
+    if value != current_field, do: broadcast_event(ctx, "update", %{"fields" => updated_fields})
+
+    {:noreply, ctx}
+  end
+
+  defp update_options(ctx, df) do
+    df = String.to_atom(df)
+    options = ctx.assigns.options[df]
+    broadcast_event(ctx, "set_axis_options", %{"options" => options})
+  end
+
+  defp to_update(field, "") do
+    %{field => nil}
+  end
+
+  defp to_update(field, value) when field in @as_int do
+    %{field => String.to_integer(value)}
+  end
+
+  defp to_update(field, value) do
+    %{field => value}
+  end
+
+  defp convert_field(field, nil), do: {String.to_atom(field), nil}
+
+  defp convert_field(field, value) when field in @as_atom do
+    {String.to_atom(field), String.to_atom(value)}
+  end
+
+  defp convert_field(field, value), do: {String.to_atom(field), value}
+
+  defp vl_alias(%Macro.Env{aliases: aliases}) do
+    case List.keyfind(aliases, VegaLite, 1) do
+      {vl_alias, _} -> vl_alias
+      nil -> VegaLite
+    end
+  end
+
+  @impl true
+  def to_attrs(ctx) do
+    ctx.assigns.fields
+    |> Map.put("vl_alias", ctx.assigns.vl_alias)
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Kino.Utils.Code.quoted_to_string()
+  end
+
+  defp to_quoted(%{"data_variable" => nil}) do
+    quote do
+    end
+  end
+
+  defp to_quoted(attrs) do
+    attrs = Map.new(attrs, fn {k, v} -> convert_field(k, v) end)
+
+    [root | nodes] = [
+      %{
+        field: nil,
+        name: :new,
+        module: attrs.vl_alias,
+        args: [[width: attrs.width, height: attrs.height]]
```

Comment:
You could do `build_arg_root(height, width)` here and simplify the build_root later on. :) similar to what we did with `arg_type`!